kani-compiler/src/kani_middle/reachability.rs

```diff
@@ -13,146 +13,135 @@
-
-    tcx.sess.abort_if_errors();
-
-    // Sort the result so code generation follows deterministic order.
-    // This helps us to debug the code, but it also provides the user a good experience since the
-    // order of the errors and warnings is stable.
-    let mut sorted_items: Vec<_> = collector.collected.into_iter().collect();
-    sorted_items.sort_by_cached_key(|item| to_fingerprint(tcx, item));
-    sorted_items
+        #[cfg(debug_assertions)]
+        collector
+            .call_graph
+            .dump_dot(tcx)
+            .unwrap_or_else(|e| tracing::error!("Failed to dump call graph: {e}"));
+
+        tcx.sess.abort_if_errors();
+
+        // Sort the result so code generation follows deterministic order.
+        // This helps us to debug the code, but it also provides the user a good experience since the
+        // order of the errors and warnings is stable.
+        let mut sorted_items: Vec<_> =
+            collector.collected.iter().map(rustc_internal::internal).collect();
+        sorted_items.sort_by_cached_key(|item| to_fingerprint(tcx, item));
+        sorted_items
+    })
+    .unwrap()
 }
 
 /// Collect all (top-level) items in the crate that matches the given predicate.
 /// An item can only be a root if they are: non-generic Fn / Static / GlobalASM
-pub fn filter_crate_items<F>(tcx: TyCtxt, predicate: F) -> Vec<MonoItem>
+pub fn filter_crate_items<F>(tcx: TyCtxt, predicate: F) -> Vec<InternalMonoItem>
 where
     F: Fn(TyCtxt, DefId) -> bool,
 {
-    let crate_items = tcx.hir_crate_items(());
-    // Filter regular items.
-    let root_items = crate_items.items().filter_map(|item| {
-        let def_id = item.owner_id.def_id.to_def_id();
-        if !is_generic(tcx, def_id) && predicate(tcx, def_id) {
-            to_mono_root(tcx, item, def_id)
-        } else {
-            None
-        }
-    });
-
-    // Filter items from implementation blocks.
-    let impl_items = crate_items.impl_items().filter_map(|impl_item| {
-        let def_id = impl_item.owner_id.def_id.to_def_id();
-        if matches!(tcx.def_kind(def_id), DefKind::AssocFn)
-            && !is_generic(tcx, def_id)
-            && predicate(tcx, def_id)
-        {
-            Some(MonoItem::Fn(Instance::mono(tcx, def_id)))
-        } else {
-            None
-        }
-    });
-    root_items.chain(impl_items).collect()
+    rustc_smir::rustc_internal::run(tcx, || {
+        let crate_items = stable_mir::all_local_items();
+        // Filter regular items.
+        crate_items
+            .iter()
+            .filter_map(|item| {
+                // Only collect monomorphic items.
+                Instance::try_from(*item).ok().and_then(|instance| {
+                    let def_id = rustc_internal::internal(item);
+                    predicate(tcx, def_id)
+                        .then_some(InternalMonoItem::Fn(rustc_internal::internal(&instance)))
+                })
```

Comment:
I'm not able to establish how this corresponds to what we had. What does `internal` do?

Reply:
This code collects all items in the current crate (`all_local_items`), and tries to instantiate them (via `Instance::try_from()`). Only the non-generic code will be successfully converted.

For those that were non-generic, we convert the stable Instance to the internal version, so we can invoke the closure that was passed as an argument which still expects the internal instance.